server/src/main/java/keywhiz/service/resources/automation/v2/BackfillRowHmacResource.java

```diff
@@ -38,11 +50,65 @@
 
   private final DSLContext jooq;
   private final RowHmacGenerator rowHmacGenerator;
+  private final AuditLog auditLog;
 
   @Inject
-  public BackfillRowHmacResource(DSLContext jooq, RowHmacGenerator rowHmacGenerator) {
+  public BackfillRowHmacResource(
+      DSLContext jooq,
+      RowHmacGenerator rowHmacGenerator,
+      AuditLog auditLog) {
     this.jooq = jooq;
     this.rowHmacGenerator = rowHmacGenerator;
+    this.auditLog = auditLog;
+  }
+
+  /**
+   * Backfill row_hmac for a specific secret.
+   */
+  @Timed @ExceptionMetered
+  @Path("secret/name/{secretName}")
+  @POST
+  @Consumes(APPLICATION_JSON)
+  @Produces(APPLICATION_JSON)
+  public void backfillSecretRowHmacByName(
+      @Auth AutomationClient automationClient,
+      @PathParam("secretName") String secretName,
+      @QueryParam("force") boolean force) {
+    Optional<SecretsRecord> maybeRow = jooq.selectFrom(SECRETS)
+        .where(SECRETS.NAME.eq(secretName))
+        .fetchOptional();
+
+    if (!maybeRow.isPresent()) {
+      throw new NotFoundException();
+    }
+
+    SecretsRecord row = maybeRow.get();
+    String oldHmac = row.getRowHmac();
+
+    if (oldHmac != null && !force) {
+      return;
+    }
+
+    String newHmac = rowHmacGenerator.computeRowHmac(
+        SECRETS.getName(),
+        List.of(row.getName(), row.getId()));
+
+    jooq.update(SECRETS)
+        .set(SECRETS.ROW_HMAC, newHmac)
+        .where(SECRETS.ID.eq(row.getId()))
+        .execute();
+
+    Map<String, String> extraInfo = ImmutableMap.<String, String>builder()
+        .put("oldHmac", Objects.toString(oldHmac))
+        .put("newHmac", newHmac)
+        .build();
+
+    auditLog.recordEvent(new Event(
+        Instant.now(),
+        EventTag.SECRET_BACKFILLHMAC,
+        automationClient.getName(),
+        secretName,
+        extraInfo));
   }
```

Comment:
Now with audit logging! Note that we use Objects.toString() so that null becomes "null" since ImmutableMap doesn't allow null values.